receiver/gitproviderreceiver/factory.go

```diff
@@ -0,0 +1,136 @@
+)
+
+// NewFactory creates a factory for the git provider receiver
+func NewFactory() receiver.Factory {
+	return receiver.NewFactory(
+		metadata.Type,
+		createDefaultConfig,
+		receiver.WithMetrics(createMetricsReceiver, metadata.MetricsStability),
+	)
+}
+
+// Gets a factory for defined scraper.
+func getScraperFactory(key string) (internal.ScraperFactory, bool) {
+	if factory, ok := scraperFactories[key]; ok {
+		return factory, true
+	}
+
+	return nil, false
+}
+
+// Create the default config based on the const(s) defined above.
+func createDefaultConfig() component.Config {
+	return &Config{
+		ScraperControllerSettings: scraperhelper.NewDefaultScraperControllerSettings(metadata.Type),
+		// TODO: metrics builder configuration may need to be in each sub scraper,
+		// TODO: for right now setting here because the metrics in this receiver will apply to all
+		// TODO: scrapers defined as a common set of gitprovider
+		// TODO: aqp completely remove these comments if the metrics build config
+		// needs to be defined in each scraper
+		// MetricsBuilderConfig: metadata.DefaultMetricsBuilderConfig(),
+	}
+}
+
+// Create the metrics receiver according to the OTEL conventions taking in the
+// context, receiver params, configuration from the component, and consumer (process or exporter)
+func createMetricsReceiver(
+	ctx context.Context,
+	params receiver.CreateSettings,
+	cfg component.Config,
+	consumer consumer.Metrics,
+) (receiver.Metrics, error) {
+
+	// check that the configuration is valid
+	conf, ok := cfg.(*Config)
+	if !ok {
+		return nil, errConfigNotValid
+	}
+
+	addScraperOpts, err := createAddScraperOpts(ctx, params, conf, scraperFactories)
+	if err != nil {
+		return nil, err
+	}
+
+	return scraperhelper.NewScraperControllerReceiver(
+		&conf.ScraperControllerSettings,
+		params,
+		consumer,
+		addScraperOpts...,
+	)
+}
+
+func createAddScraperOpts(
+	ctx context.Context,
+	params receiver.CreateSettings,
+	cfg *Config,
+	factories map[string]internal.ScraperFactory,
+) ([]scraperhelper.ScraperControllerOption, error) {
+	scraperControllerOptions := make([]scraperhelper.ScraperControllerOption, 0, len(cfg.Scrapers))
+
+	for key, cfg := range cfg.Scrapers {
+		gitProviderScraper, ok, err := createGitProviderScraper(ctx, params, key, cfg, factories)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create scraper %q: %w", key, err)
+		}
+
+		if ok {
+			scraperControllerOptions = append(scraperControllerOptions, scraperhelper.AddScraper(gitProviderScraper))
+			continue
+		}
+
+		return nil, fmt.Errorf("git provider scraper factory not found for key: %q", key)
+
+	}
+
+	return scraperControllerOptions, nil
+}
+
+func createGitProviderScraper(
+	ctx context.Context,
+	params receiver.CreateSettings,
+	key string,
+	cfg internal.Config,
+	factories map[string]internal.ScraperFactory,
+) (scraper scraperhelper.Scraper, ok bool, err error) {
```

Comment:
Why do you need both a "ok" flag and an error? If the error is not nil, isn't it automatically not ok? Are there scenarios where OK is true and error is not nil, or that OK is false and the error is nil?

Reply:
Yes, there are scenarios. This was mimic'd from the hostmetricsreceiver as well during the transition to multiple scrapers instead of just one. The logic here is that the calling function will check whether it's an error from creating a scraper or if the scraper just doesn't exist, and return an error message accordingly. 

The scraper could exist, but fail to create (that would be err) or it could not be ok, in that the scraper isn't a actual scraper. 

We could refactor this such that the error message is just returned from the `createGitProviderScraper` directly up, thereby not needing to do the error & ok checking at the calling function. From a troubleshooting perspective, I don't think it's important to log both possibilities though.

Reply:
> The logic here is that the calling function will check whether it's an error from creating a scraper or if the scraper just doesn't exist, and return an error message accordingly

There could be two different errors, and the caller here would just handle the errors accordingly?

Reply:
updated the logic to remove `ok` entirely